x-pack/dockerlogbeat/pipelinemanager/config.go

```diff
@@ -0,0 +1,142 @@
+	TemplateName    string `struct:"setup.template.name,omitempty"`
+	TempatePattern  string `struct:"setup.template.pattern,omitempty"`
+}
+
+// NewCfgFromRaw returns a ContainerOutputConfig based on a raw config we get from the API
+func NewCfgFromRaw(input map[string]string) (ContainerOutputConfig, error) {
+
+	newCfg := ContainerOutputConfig{}
+	endpoint, ok := input["endpoint"]
+	if !ok {
+		return newCfg, errors.New("An endpoint flag is required")
+	}
+	newCfg.Endpoint = endpoint
+
+	var isIndex bool
+
+	newCfg.User = input["user"]
+	newCfg.Password = input["password"]
+	newCfg.Index, isIndex = input["index"]
+	newCfg.Pipeline = input["pipeline"]
+	newCfg.CloudID = input["cloud_id"]
+	newCfg.CloudAuth = input["cloud_auth"]
+	newCfg.ProxyURL = input["proxy_url"]
+	newCfg.APIKey = input["api_key"]
+	newCfg.Timeout = input["timeout"]
+	newCfg.BackoffInit = input["backoff_init"]
+	newCfg.BackoffMax = input["backoff_max"]
+
+	rawILM, isILM := input["ilm_enabled"]
+
+	if isILM {
+		ilmBool, err := strconv.ParseBool(rawILM)
+		if err != nil {
+			return newCfg, errors.Wrapf(err, "could not convert %s to bool", rawILM)
+		}
+		newCfg.ILMEnabled = ilmBool
+
+		if isIndex && newCfg.ILMEnabled {
+			return newCfg, errors.New("Cannot set output index while ILM is enabled")
+		}
+
+	}
+
+	newCfg.ILMRollverAlias = input["ilm_rollover_alias"]
+	newCfg.ILMPatterns = input["ilm_pattern"]
+
+	if isIndex {
+		tname, tnameOk := input["template_name"]
+		tpattern, tpatternOk := input["template_pattern"]
+
+		if !tnameOk || !tpatternOk {
+			return newCfg, errors.New("template_pattern and template_name must be set if index is set")
+		}
+		newCfg.TempatePattern = tpattern
+		newCfg.TemplateName = tname
+	}
+
+	return newCfg, nil
+}
+
+// CreateConfig converts the struct into a config object that can be absorbed by libbeat
+func (cfg ContainerOutputConfig) CreateConfig() (*common.Config, error) {
+
+	// the use of typeconv is a hacky shim so we can impliment `omitempty` where needed.
+	var tmp map[string]interface{}
+	err := typeconv.Convert(&tmp, cfg)
+	if err != nil {
+		return nil, errors.Wrap(err, "error converting config struct to interface")
+	}
+	cfgFinal, err := common.NewConfigFrom(tmp)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating config object")
+	}
+
+	return cfgFinal, nil
+}
+
+// GetHash returns a sha1 hash of the config
+func (cfg ContainerOutputConfig) GetHash() string {
+	var hashString string
+	var orderedVal []string
+
+	values := reflect.ValueOf(cfg)
+	for i := 0; i < values.NumField(); i++ {
+		valRaw := values.Field(i).Interface()
+		if parsedVal, ok := valRaw.(string); ok {
+			orderedVal = append(orderedVal, parsedVal)
+		} else if parsedVal, ok := valRaw.(bool); ok {
+			if parsedVal {
+				orderedVal = append(orderedVal, "true")
+			} else {
+				orderedVal = append(orderedVal, "false")
+			}
+		}
+
+	}
+
+	sort.Strings(orderedVal)
+
+	for _, val := range orderedVal {
+		hashString = hashString + val
+	}
```

Comment:
the `Hash` interface implements `io.Writer` as well. Consider `bytes.Buffer` or use `io.WriteString(hasher, val)` to reduce allocations.